### PR TITLE
`Development`:  Refactor server integration tests for notifications

### PR DESCRIPTION
**Pull request by Malyuk-A**
#### General
- This PR refactors server tests so the only testing besides code review is executing the tests locally
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.

### Motivation and Context
@krusche asked me to refactor the server tests in `java/de/tum/in/www1/artemis/service/notifications` as my last PR.
These tests were focused on testing the isolated functionality (~UnitTest).
Now they are more like Integration Tests, i.e. they actually invoke the real Services and Repositories instead of Mocks.

### Steps for Testing
This PR refactors the server tests for notifications, 
thus the only possible testing is looking at the code and executing the tests locally.

#### Code Review
- [x] Review 1
- [x] Review 2

### Test Coverage
GroupNotificationService = 100%
SingleUserNotificationService ~ 100% 
NotificationSettingsService = 100%
NotificationTargetFactory ~ 100%
(Via real calls -> MailService all related functionality for notifications based emails)

### CODE REVIEW

File: src/test/java/de/tum/in/www1/artemis/service/notifications/GroupNotificationServiceTest.java [hunk 1]
```diff
@@ -117,95 +96,61 @@
...
+        ExerciseGroup exerciseGroup = new ExerciseGroup();
         exerciseGroup.setExam(exam);
 
         examExercise = new TextExercise();
         examExercise.setExerciseGroup(exerciseGroup);
         examExercise.setProblemStatement(EXAM_PROBLEM_STATEMENT);
 
-        quizExercise = new QuizExercise();
-        quizExercise.setCourse(course);
+        quizExercise = database.createQuiz(course, null, null);
```
Comment (willergeorg) [inline src/test/java/de/tum/in/www1/artemis/service/notifications/GroupNotificationServiceTest.java:133]: I'm adding the comment here, but I'm actually referring to line 129-131 in your changes. Is there a particular reason you decided not to use `Modelfactory.generateTextExerciseForExam()`?

File: src/test/java/de/tum/in/www1/artemis/service/notifications/SingleUserNotificationServiceTest.java [hunk 2]
```diff
@@ -1,109 +1,72 @@
...
-        singleUserNotificationRepository = mock(SingleUserNotificationRepository.class);
+        course = database.createCourse();
 
-        singleUserNotificationService = spy(
-                new SingleUserNotificationService(singleUserNotificationRepository, userRepository, messagingTemplate, mailService, notificationSettingsService));
+        List<User> users = database.addUsers(1, 0, 0, 0);
+        user = users.get(0);
 
-        exercise = new TextExercise();
+        Exercise exercise = new TextExercise();
```
Comment (willergeorg) [inline src/test/java/de/tum/in/www1/artemis/service/notifications/SingleUserNotificationServiceTest.java:63]: `Modelfactory.generateTextExercise()` would also be considerable here

File: src/test/java/de/tum/in/www1/artemis/service/notifications/SingleUserNotificationServiceTest.java [hunk 3]
```diff
@@ -1,109 +1,72 @@
...
         exercise.setCourse(course);
 
         fileUploadExercise = new FileUploadExercise();
         fileUploadExercise.setCourse(course);
 
-        lecture = new Lecture();
+        Lecture lecture = new Lecture();
```
Comment (willergeorg) [inline src/test/java/de/tum/in/www1/artemis/service/notifications/SingleUserNotificationServiceTest.java:69]: `Modelfactory.generateLecture()` would also be considerable here

**Review by willergeorg — COMMENTED** [review]
I had a look at your code. You taught me everything I currently know about our server tests, nevertheless I marked a few minor suggestions/questions. They are not critical in your case as your current implementation is not much longer. Tell me what you think about it. 🙏

File: src/test/java/de/tum/in/www1/artemis/service/notifications/GroupNotificationServiceTest.java (hunk 1 above)
Comment (Malyuk-A) [reply src/test/java/de/tum/in/www1/artemis/service/notifications/GroupNotificationServiceTest.java:133]: Sure, good question!
From my understanding (and the fact that we have no real convention on how to do this (no information about this in our documentation)) this is basically the same thing as long as you do not mock it.
We prefer to use real entity-type objects and how you create them is not as important.
E.g. in this case the NotificationServices use e.g. exercises as a transient object (it is not pulled or saved/changed to the DB), i.e. it works just as any other input. Therefore, It is enough to have the needed properties set and not a lot more than the database function provide
image
image

File: src/test/java/de/tum/in/www1/artemis/service/notifications/SingleUserNotificationServiceTest.java (hunk 2 above)
Comment (Malyuk-A) [reply src/test/java/de/tum/in/www1/artemis/service/notifications/SingleUserNotificationServiceTest.java:63]: have a look at my first response

File: src/test/java/de/tum/in/www1/artemis/service/notifications/SingleUserNotificationServiceTest.java (hunk 3 above)
Comment (Malyuk-A) [reply src/test/java/de/tum/in/www1/artemis/service/notifications/SingleUserNotificationServiceTest.java:69]: have a look at my first response

File: src/test/java/de/tum/in/www1/artemis/service/notifications/GroupNotificationServiceTest.java (hunk 1 above)
Comment (willergeorg) [reply src/test/java/de/tum/in/www1/artemis/service/notifications/GroupNotificationServiceTest.java:133]: I understand. As I already said, in your case the use of the `Modelfactory` would not simplify the code critically imo. Therefore your way is fine for me.

**Review by willergeorg — APPROVED** [review]
After discussing some spots with Alex, the code looks reasonable for me.

**Review by filip-gregurevic — APPROVED** [review]
The tests look good to me, nice job 👍🏼

**Pull request merged**